3d export stl

ptitjano opened the pull request:
## Description

This adds support to export a 3D scene in the [STL](https://en.wikipedia.org/wiki/STL_(file_format)) format in addition to the existing `OBJ` one. The `STL` format is much simpler than the `OBJ`. For example, it does not handle textures.

This new workflow uses the exist export scene logic. The change relies mainly on 3 changes:
- Add an `ExportFormat` enum to choose the export format
- the `STL`  logic is added in `Qgs3DExportObject` 
- Adapt `QgsMap3DExportWidget` to be able to chosse between `STL` and `OBJ` format

## new Export widget

<img width="896" height="574" alt="image" src="https://github.com/user-attachments/assets/5599b822-5f12-4548-9cb3-c9a55aaa75db" />

## update 9 February

Added a STL info message to warn about 3D printing limitations

<img width="1038" height="650" alt="Capture d’écran du 2026-02-09 17-51-29" src="https://github.com/user-attachments/assets/c2a63232-a696-4d20-a673-63ad3259d23e" />



**Funded by Stadt Frankfurt am Main**

cc @benoitdm-oslandia

ptitjano committed:
qgis: Add an 3D export format enum

This will be used in the following commits to add a stl format export

ptitjano committed:
qgs3dsceneexporter: Add support for stl export format

The existing `Qgs3DSceneExporter::save` function has a new parameter
`exportFormat` to choose between `Stl` and `Obj`. It is split in two
functions `saveObj` which saves in Obj format and reuses the
existing code from `save` and the new `saveStl` function which saves
an object in stl format.

The new `saveStl` is similar to the existing `saveObj` one. One
iterate over the different objects of the scene and call
`Qgs3dExportObject::saveTo`. The `stl` logic is simpler to the the
`obj` one because, `stl` cannot save textures.

ptitjano committed:
qgs3dsceneexporter: Factor out logic to compute scene center and scale

This will be used by the `Stl` and the `Obj` exporter in the next
commit.

ptitjano committed:
qgs3dexportobject: Add stl export format

The existing `Qgs3DExportObject::saveTo` function has a new parameter
`exportFormat` to choose between `Stl` and `Obj`. It is split in two
functions `saveToObj` which saves in Obj format and reuses the
existing code from `saveTo` and the new `saveToStl` function which
saves an object in stl format. The `stl` format is simpler than the
`obj` one. One can not save a texture and one cannot smooth edges and
save normals.

ptitjano committed:
qgs3dmapexportsettings: Add support for ExportFormat

This setting in `Qgs3DMapScene::exportScene` to export in `obj` or
`stl` format.

ptitjano committed:
qgsmap3dexportwidget: Add option to choose between obj and stl format

ptitjano committed:
qgs3dmapexportsettings: Add a exportFileUri method

This will make it easier to handle `Obj` and `Stl` formats.

ptitjano committed:
qgs3dexportobject: Fix clang-tidy warning

ptitjano committed:
testqgs3dutils: Move testExportToObj to testqgs3dexporter

This makes more sense. `Qgs3DExportObject` is not part of `Qgs3DUtils`.

ptitjano committed:
qgs3dsceneexporter: Directly save object name in Qgs3DExportObject

The current `save` logic in `Qgs3DSceneExporter` iterates overs the
differents objects, save the material, sets the object name, sets the
material name, and finally saves the object geometry.

This means that one cannot directly use `Qgs3DExportObject` to save an
object. One need to save the object name and then call
`Qgs3DExportObject::saveTo`. Besides, the object name is part of the
save object logic.

This current logic is simplified by moving the object name and
material name logic inside `Qgs3DExportObject::saveTo`.

ptitjano committed:
qgsmap3dexportwidget: Rename it to qgs3dmapexportwidget

This way, the name aligns aligns with the names of the other files.

benoitdm-oslandia reviewed (commented):
Great! Just a bit of more doc please!

nyalldawson commented on src/3d/qgs3dmapexportsettings.cpp:
```diff
@@ -41,3 +43,10 @@ Qgs3DMapExportSettings::~Qgs3DMapExportSettings()
   settings.setValue( QStringLiteral( "UI/last3DSceneExportExportTextures" ), mExportTextures );
 }
+
+QString Qgs3DMapExportSettings::exportFilePath() const
```
can you rename to exportFileUrl instead, as that's what it's actually returning?

nyalldawson commented on tests/src/3d/testqgs3dexporter.cpp:
```diff
@@ -86,7 +91,194 @@ void TestQgs3DExporter::cleanupTestCase()
+
+  // case where all vertices are used
+  {
+    const QString object_name = QStringLiteral( "all_faces" );
+    const QString object_filename = object_name + QStringLiteral( "." ) + exportExtension;
```
naming convention

nyalldawson commented on tests/src/3d/testqgs3dexporter.cpp:
```diff
@@ -86,7 +91,194 @@ void TestQgs3DExporter::cleanupTestCase()
+      // 27, 28, 29,
+    };
+
+    const QString object_name = QStringLiteral( "sparse_faces" );
+    const QString object_filename = object_name + QStringLiteral( "." ) + exportExtension;
```
naming convention

nyalldawson commented on tests/src/3d/testqgs3dexporter.cpp:
```diff
@@ -177,16 +370,19 @@ void TestQgs3DExporter::test3DSceneExporter()
-  do3DSceneExport( "scene_export", 4, 3, nbFeat, nbFaces, scene, layerPoly, &engine );
-  // =========== check with 25 tiles ==> 3 exported objects
-  do3DSceneExport( "scene_export", 5, 3, nbFeat, nbFaces, scene, layerPoly, &engine );
+  for ( const auto &exportFormat : { Qgs3DTypes::ExportFormat::Obj, Qgs3DTypes::ExportFormat::Stl } )
```
no auto here

nyalldawson commented on tests/src/3d/testqgs3dexporter.cpp:
```diff
@@ -235,16 +431,19 @@ void TestQgs3DExporter::test3DSceneExporterBig()
-  do3DSceneExport( "big_scene_export", 4, 32, nbFeat, nbFaces, scene, mLayerBuildings, &engine );
-  // =========== check with 25 tiles ==> 70 exported objects
-  do3DSceneExport( "big_scene_export", 5, 70, nbFeat, nbFaces, scene, mLayerBuildings, &engine );
+  for ( const auto &exportFormat : { Qgs3DTypes::ExportFormat::Obj, Qgs3DTypes::ExportFormat::Stl } )
```
no auto here

ptitjano replied:
Done.

nyalldawson commented on tests/src/3d/testqgs3dexporter.cpp:
```diff
@@ -292,7 +491,10 @@ void TestQgs3DExporter::test3DSceneExporterFlatTerrain()
   scene->cameraController()->setLookingAtPoint( QVector3D( 0, 0, 0 ), 1500, 40.0, -10.0 );
 
-  do3DSceneExport( "flat_terrain_scene_export", 5, 70, 401, 19875, scene, mLayerBuildings, &engine, scene->terrainEntity() );
+  for ( const auto &exportFormat : { Qgs3DTypes::ExportFormat::Obj, Qgs3DTypes::ExportFormat::Stl } )
```
no auto here

nyalldawson commented:
I'd also like to re-state my previous objection to trying to compare test strings against multiple 40k+ line files. These tests will be impossible to debug in future.

ptitjano commented on python/3d/auto_generated/qgs3dtypes.sip.in:
```diff
@@ -40,6 +43,12 @@ Defines enumerations and other auxiliary types for QGIS 3D.
       Selected3DRenderer,
     };
+
+    enum class ExportFormat
```
Done.

ptitjano commented on src/3d/qgs3dmapexportsettings.h:
```diff
@@ -52,6 +53,7 @@ class _3D_EXPORT Qgs3DMapExportSettings
     int terrainTextureResolution() const { return mTerrainTextureResolution; }
     //! Returns the scale of the exported model
     float scale() const { return mScale; }
+    Qgs3DTypes::ExportFormat exportFormat() const { return mExportFormat; }
```
Done.

ptitjano commented on src/app/3d/qgs3dmapexportwidget.cpp:
```diff
@@ -78,9 +84,23 @@ void QgsMap3DExportWidget::settingsChanged()
   mExportSettings->setExportTextures( ui->exportTexturesCheckBox->isChecked() );
   mExportSettings->setTerrainTextureResolution( ui->terrainTextureResolutionSpinBox->value() );
   mExportSettings->setScale( ui->scaleSpinBox->value() );
+  mExportSettings->setExportFormat( Qgs3DTypes::ExportFormat( ui->exportFormatComboxBox->currentData().toInt() ) );
```
Done.

ptitjano commented on tests/src/3d/testqgs3dexporter.cpp:
```diff
@@ -86,7 +91,194 @@ void TestQgs3DExporter::cleanupTestCase()
+    0.676449, 0, -0.736489
+  };
+
+  const QString myTmpDir = QDir::tempPath() + '/';
```
Done.

ptitjano commented:
> I'd also like to re-state my previous objection to trying to compare test strings against multiple 40k+ line files. These tests will be impossible to debug in future.

I will re-state my previous answer: https://github.com/qgis/QGIS/pull/63173#issuecomment-3312798162
@benoitdm-oslandia and I strongly believe that those tests and files are useful (necessary) to debug and solve export issues. Both of us have spent a lot of time debugging and fixing regressions on the 3d scene export.

Anyway, as a compromise, I have removed the big stl export terrain. The main idea was to easily detect a regression on terrain export. This is already covered by the obj test.

benoitdm-oslandia commented:
Since there is no news on this PR, it will be merged next week. Regards.

uclaros reviewed (changes requested):
What's the rationale behind exporting to the very inefficient ascii variant of the stl format and not the more commonly used binary one?

<!-- osl:870b1a90550 -->

ptitjano committed:
qgs3dexportobject: Use std::as_const instead of qAsConst

ptitjano committed:
qgs3dsceneexporter: Use std::as_const instead of qAsConst

ptitjano commented on src/core/qgis.h:
```diff
@@ -4147,6 +4147,18 @@ class CORE_EXPORT Qgis
+     *
+     * \since QGIS 4.0
+     */
+    enum class ExportFormat : int
```
changed

ptitjano commented on src/3d/qgs3dsceneexporter.cpp:
```diff
@@ -876,18 +875,71 @@ bool Qgs3DSceneExporter::save( const QString &sceneName, const QString &sceneFol
+  float maxX = minFloat;
+  float maxY = minFloat;
+  float maxZ = minFloat;
+  for ( Qgs3DExportObject *obj : qAsConst( mObjects ) )
```
done.

ptitjano committed:
qgs3dmapexportwidget: Add info message regarding stl export

nyalldawson commented:
@ptitjano 

> Thanks for your feedback. Indeed, STL is usually used for 3d printing. However, I still think that it can be exposed in the GUI for the following reasons.
> vector layers can now be closed with the new "Walls, Roof and Floor" option in the 3d symbology
> with https://github.com/qgis/QGIS/pull/64491 merged, terrain export can be disabled. This means, that a scene with closed > objects can be disabled
> Also, I have added a "infos" box to explain current usage and limitations.


What's the purpose of just exporting an STL with isolated/floating buildings (and no terrain) though? Unless every building was sitting on a flat plain, it's going to be basically meaningless to print (you'd just get a model with buildings held up to different heights by print supports, and I can't see why anyone would want that).  And if you DID want all buildings on a flat plain, you'd still have to do post-processing edits to ADD that plane before you could slice the model. So the only viable use case I see for that would be 3d printing a SINGLE building.

Please correct me if I'm missing something here!

ptitjano commented:
> @ptitjano
> 
> > Thanks for your feedback. Indeed, STL is usually used for 3d printing. However, I still think that it can be exposed in the GUI for the following reasons.
> > vector layers can now be closed with the new "Walls, Roof and Floor" option in the 3d symbology
> > with #64491 merged, terrain export can be disabled. This means, that a scene with closed > objects can be disabled
> > Also, I have added a "infos" box to explain current usage and limitations.
> 
> What's the purpose of just exporting an STL with isolated/floating buildings (and no terrain) though? Unless every building was sitting on a flat plain, it's going to be basically meaningless to print (you'd just get a model with buildings held up to different heights by print supports, and I can't see why anyone would want that). And if you DID want all buildings on a flat plain, you'd still have to do post-processing edits to ADD that plane before you could slice the model. So the only viable use case I see for that would be 3d printing a SINGLE building.
> 
> Please correct me if I'm missing something here!

You're right. For now, the only case where you can be sure the generated file is directly printable is when there is a single object. As you previously mentioned, I also agree that a user who exports a file and expects to print it immediately might be disappointed.

However, my main concern is that if we don’t expose this feature in the GUI, no one will use it because they won’t know it exists. Not showing the feature is basically the same as making it unused.

Downloading an STL model and editing it before printing is a common step in many 3D printing workflows. With the optional terrain export, users can choose to use the terrain to close the scene, or not export it and create their own closure.

My compromise would be to rephrase the information message to clearly state that the exported scene is not closed and must be edited before printing.

This way, users know the feature exists while also understanding that they need to modify the export before 3D printing. This allows users to easily export STL files while keeping expectations clear

signedav commented:
I'm not familiar enough with the subject. But out of curiosity:  Isn't there a use case where you would want the terrain exported as an unclosed STL @ptitjano? For example, to use it for an animation in blender or in other software that does not support OBJ?

nyalldawson commented:
@ptitjano I've decided I'll remove my previous objections to exposing this. If your client is happy, then let's go for it. We don't owe anything to QGIS users who aren't funding development. :+1:

uclaros commented:
> We don't owe anything to QGIS users who aren't funding development. 👍

@nyalldawson I agree on unblocking this, can you please elaborate on the above though? Because IMHO, blindly following funders' requests is a recipe for disaster.

ptitjano commented:
rebased and fixed conflicts.

ptitjano committed:
qgstest: Fix clang-tidy warning in checkLongStr

Replace new char[] with qstrdup() to match the free() call expected by
QTest::compare_helper.

benoitdm-oslandia commented:
As all points have been resolved. I will merge

benoitdm-oslandia commented:
@uclaros can you remove your "requested change" status please? Or I can do it if you want to?
